File: Algorithms/33. Search in Rotated Sorted Array.rs
// brute force
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let mut i = 0;
        while i < nums.len() {
            if nums[i] == target { break; }
            else { i += 1; }
        }
        
        if i == nums.len() { return -1; }
        i as i32
    }
}

// binary search
impl Solution2 {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let mut l = 0;
        let mut r = nums.len() - 1;
        while l <= r {
            let mid = l + r >> 1;
            if nums[mid] == target {
                return mid as i32;
            }

            if nums[l] <= nums[mid] {
                if target < nums[mid] && target >= nums[l] {
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            } else {
                if target > nums[mid] && target <= nums[r] {
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            }
        }
        -1
    }
}

File: Algorithms/33. Search in Rotated Sorted Array.ts
function search(nums: number[], target: number): number {
  let l: number = 0;
  let r: number = nums.length - 1;
  while (l <= r) {
    const mid: number = Math.floor((l + r) / 2);
    if (nums[mid] === target) {
      return mid;
    }
    if (nums[mid] < nums[r]) {
      if (nums[mid] < target && nums[r] >= target) {
        l = mid + 1;
      } else {
        r = mid - 1;
      }
    } else {
      if (nums[l] <= target && nums[mid] > target) {
        r = mid - 1;
      } else {
        l = mid + 1;
      }
    }
  }
  return -1;
}
